src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,559 @@
+
+std::string even_bytes(const std::string& a) {
+  std::stringstream ss;
+  size_t len = a.length();
+  for (size_t j = 0; j < len; j += 2) {
+    ss << a[j];
+  }
+  return ss.str();
+}
+
+std::string hash_two_strings(const std::string& a, const std::string& b) {
+  bool a_shorter = (a.length() < b.length());
+  const std::string* shorter = (a_shorter ? &a : &b);
+  const std::string* longer  = (a_shorter ? &b : &a);
+  uint64_t hashA = std::hash<std::string>()(*longer);
+  uint64_t hashB = std::hash<std::string>()(*shorter + even_bytes(*longer));
+  if (sizeof(size_t) < 8) {
+    // On a 32bit machine we need to work harder to get 80 bits
+    uint64_t hashC = std::hash<std::string>()(*longer + "SmallChangeForThirdHash");
+    hashA += (hashC << 32);
+  }
+  if (a_shorter) hashA ^= 0xff; // order of strings should change result
+  hashA ^= ALL_HEADER_CKSUM; // checksum from all header files
+  hashA ^= (0x3ac5*hashB); // mix low bits of hashB into hashA
+  hashB &= 0xffff;
+  char buf[128];
+  char* out = buf;
+  for (int pos = 80; (pos -= 4) >= 0;) {
+    int64_t nibble = ((pos >= 64) ? (hashB >> (pos-64)) : (hashA >> pos)) & 0xf;
+    *out++ = ((nibble < 10) ? nibble+'0' : nibble-10+'a');
+  }
+  *out = 0;
+  return std::string(buf);
+}
+
+bool file_exists(const std::string& name) {
+  struct stat st;
+  int rc = ::stat(name.c_str(), &st);
+  return (rc == 0);
+}
+
+std::string read_file_as_string(const std::string& name) {
+  FILE* f = fopen(name.c_str(), "r");
+  if (!f) return std::string("");
+  std::stringstream ss;
+  for (;;) {
+    int c = fgetc(f);
+    if (c == EOF) break;
+    ss << (char)c;
+  }
+  fclose(f);
+  return ss.str();
+}
+
+std::string get_module_dir() {
+  // This gives us a distinct temp directory for each process, so that
+  // we can manage synchronization of threads accessing files in
+  // the temp directory using only the in-memory big_mutex, rather than
+  // dealing with the complexity of file locking.
+  char buf[512];
+  strcpy(buf, "/tmp/hail_XXXXXX");
+  return ::mkdtemp(buf);
+}
+
+class ModuleConfig {
+ public:
+  bool is_darwin_;
+  std::string java_md_;
+  std::string ext_cpp_;
+  std::string ext_lib_;
+  std::string ext_mak_;
+  std::string module_dir_;
+
+ public:
+  ModuleConfig() :
+#if defined(__APPLE__) && defined(__MACH__)
+    is_darwin_(true),
+#else
+    is_darwin_(false),
+#endif
+    java_md_(is_darwin_ ? "darwin" : "linux"),
+    ext_cpp_(".cpp"),
+    ext_lib_(is_darwin_ ? ".dylib" : ".so"),
+    ext_mak_(".mak"),
+    module_dir_(get_module_dir()) {
+  }
+  
+  std::string get_lib_name(const std::string& key) {
+    std:: stringstream ss;
+    ss << module_dir_ << "/hm_" << key << ext_lib_;
+    return ss.str();
+  }
+  
+  void ensure_module_dir_exists() {
+    int rc = ::access(module_dir_.c_str(), R_OK);
+    if (rc < 0) { // create it
+      rc = ::mkdir(module_dir_.c_str(), 0666);
+      if (rc < 0) perror(module_dir_.c_str());
+      rc = ::chmod(module_dir_.c_str(), 0755);
```

Comment:
Curious, why are the chmods necessary?

Reply:
We could take those out, on the assumption that the files are per-session and thus will never be read or written by someone with a different uid.  Though with a minor concern that it might affect debugging if the Spark code was running as some other user, and we then wanted to look at the files afterwards to see what went wrong.

Reply:
Ah, OK.  I'm not aware that this would be necessary on any machine/setup we've run on, but fine with me if you leave it in, either way.